File: src/java/src/main/java/org/elastos/hive/FileStorage.java
package org.elastos.hive;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;

class FileStorage implements DataStorage {
	private static final String BACKUP = "credential-backup";
	private static final String TOKENS = "tokens";

	private String basePath;

	public FileStorage(String rootPath, String userDid) {
		String path = rootPath;
		if (!path.endsWith(File.separator))
			path += File.separator;

		path += compatDid(userDid);
		this.basePath = path;

		File file = new File(path);
		if (!file.exists())
			file.mkdirs();
	}

	@Override
	public String loadBackupCredential(String serviceDid) {
		return readContent(makeFullPath(BACKUP, compatDid(serviceDid)));
	}

	@Override
	public String loadAccessToken(String serviceDid) {
		return readContent(makeFullPath(TOKENS, compatDid(serviceDid)));
	}

	@Override
	public String loadAccessTokenByAddress(String providerAddress) {
		return readContent(makeFullPath(TOKENS, sha256(providerAddress)));
	}

	@Override
	public void storeBackupCredential(String serviceDid, String credential) {
		writeContent(makeFullPath(BACKUP, compatDid(serviceDid)), credential);
	}

	@Override
	public void storeAccessToken(String serviceDid, String accessToken) {
		writeContent(makeFullPath(TOKENS, compatDid(serviceDid)), accessToken);
	}

	@Override
	public void storeAccessTokenByAddress(String providerAddress, String accessToken) {
		writeContent(makeFullPath(TOKENS, sha256(providerAddress)), accessToken);
	}

	@Override
	public void clearBackupCredential(String serviceDid) {
		deleteContent(makeFullPath(BACKUP, compatDid(serviceDid)));
	}

	@Override
	public void clearAccessToken(String serviceDid) {
		deleteContent(makeFullPath(TOKENS, compatDid(serviceDid)));
	}

	@Override
	public void clearAccessTokenByAddress(String providerAddress) {
		deleteContent(makeFullPath(TOKENS, sha256(providerAddress)));
	}

	private String readContent(String path) {
		if (path == null)
			return null;

		Path p = Paths.get(path);
		if (!Files.exists(p))
			return null;

		try {
			return new String(Files.readAllBytes(p));
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}
	}

	private void writeContent(String path, String content) {
		if (path == null)
			return;

		File parent = Paths.get(path).getParent().toFile();
		if (!parent.exists())
			parent.mkdirs();

		if (!parent.exists())
			return;

		try {
			Files.write(Paths.get(path), content.getBytes(StandardCharsets.UTF_8));
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	private void deleteContent(String path) {
		if (path == null)
			return;

		try {
			Files.deleteIfExists(Paths.get(path));
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	private String makeFullPath(String segPath, String fileName) {
		return this.basePath
					+ File.separator
					+ segPath
					+ File.separator
					+ fileName;
	}

	private String compatDid(String did) {
		String[] parts = did.split(":");
		return parts.length >= 3 ? parts[2] : did;
	}

	private String sha256(String message) {
		byte[] bytes;

		try {
			MessageDigest digest = MessageDigest.getInstance("SHA-256");
			digest.update(message.getBytes("UTF-8"));
			bytes = digest.digest();
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		}

		StringBuffer buffer = new StringBuffer();
		String temp = null;

		for (int i = 0; i < bytes.length; i++) {
			temp = Integer.toHexString(bytes[i] & 0xFF);
			if (temp.length() == 1)
				buffer.append("0");

			buffer.append(temp);
		}
		return buffer.toString();
	}
}

File: src/domain/filestorage.ts
import { File } from './file'
import { DataStorage } from './datastorage'
import { SHA256 } from './sha256'
import { BASE64 } from './base64'

export class FileStorage implements DataStorage {
	private static BACKUP = "credential-backup";
	private static TOKENS = "tokens";

	private basePath: string;

	constructor(rootPath: string, userDid: string) {
		let path = rootPath;
		if (!path.endsWith(File.SEPARATOR))
			path += File.SEPARATOR;

		path += this.compatDid(userDid);
		this.basePath = path;

		let file = new File(path);
		if (!file.exists())
			file.createDirectory();
	}

	public loadBackupCredential(serviceDid: string): string {
		return this.readContent(this.makeFullPath(FileStorage.BACKUP, this.compatDid(serviceDid)));
	}

	public loadAccessToken(serviceDid: string): string {
		return this.readContent(this.makeFullPath(FileStorage.TOKENS, this.compatDid(serviceDid)));
	}

	public loadAccessTokenByAddress(providerAddress: string): string {
		return this.readContent(this.makeFullPath(FileStorage.TOKENS, SHA256.encodeToString(Buffer.from(providerAddress))));
	}

	public storeBackupCredential(serviceDid: string, credential: string): void {
		this.writeContent(this.makeFullPath(FileStorage.BACKUP, this.compatDid(serviceDid)), credential);
	}

	public storeAccessToken(serviceDid: string, accessToken: string): void {
		this.writeContent(this.makeFullPath(FileStorage.TOKENS, this.compatDid(serviceDid)), accessToken);
	}

	public storeAccessTokenByAddress(providerAddress: string, accessToken: string): void {
		this.writeContent(this.makeFullPath(FileStorage.TOKENS, SHA256.encodeToString(Buffer.from(providerAddress))), accessToken);
	}

	public clearBackupCredential(serviceDid: string): void {
		this.deleteContent(this.makeFullPath(FileStorage.BACKUP, this.compatDid(serviceDid)));
	}

	public clearAccessToken(serviceDid: string): void {
		this.deleteContent(this.makeFullPath(FileStorage.TOKENS, this.compatDid(serviceDid)));
	}

	public clearAccessTokenByAddress(providerAddress: string): void {
		this.deleteContent(this.makeFullPath(FileStorage.TOKENS, SHA256.encodeToString(Buffer.from(providerAddress))));
	}

	private readContent(path: string): string {
		if (path == null)
			return null;

		if (!File.exists(path))
			return null;

		try {
			return new File(path).readText();
		} catch (e) {
            console.log(e.message);
			return null;
		}
	}

	private writeContent(path: string, content: string): void {
		if (path == null)
			return;

        let targetFile: File = new File(path)
		let parent: File = targetFile.getParentDirectory();
		if (!parent.exists())
			parent.createDirectory();

		if (!parent.exists())
			return;

		try {
            targetFile.writeText(content);
		} catch (e) {
			console.log(e.message);
		}
	}

	private deleteContent(path: string): void {
		if (path == null)
			return;

		try {
            (new File(path)).delete();
		} catch (e) {
			console.log(e.message);
		}
	}

	private makeFullPath(segPath: string, fileName: string): string {
		return this.basePath
					+ File.SEPARATOR
					+ segPath
					+ File.SEPARATOR
					+ fileName;
	}

	private compatDid(did: string): string {
		let parts: Array<string> = did.split(":");
		return parts.length >= 3 ? parts[2] : did;
	}
}
